polygeist/lib/Conversion/PolygeistToLLVM/PolygeistToLLVM.cpp

```diff
@@ -370,6 +370,32 @@ struct Pointer2MemrefOpLowering
   }
 };
 
+struct AddrSpaceCastOpLowering
+    : public ConvertOpToLLVMPattern<AddrSpaceCastOp> {
+  using ConvertOpToLLVMPattern<AddrSpaceCastOp>::ConvertOpToLLVMPattern;
+
+  LogicalResult
+  matchAndRewrite(AddrSpaceCastOp op, OpAdaptor transformed,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto loc = op.getLoc();
+    MemRefType resTy = op.getType();
+    MemRefType srcTy = op.getSource().getType();
+    assert(srcTy.getElementType() == resTy.getElementType() &&
+           "Expecting same source and result element type");
+    Type elemTy = resTy.getElementType();
+    Value ptr = rewriter.create<Memref2PointerOp>(
+        loc, LLVM::LLVMPointerType::get(elemTy, srcTy.getMemorySpaceAsInt()),
+        op.getSource());
+    ptr = rewriter.create<LLVM::AddrSpaceCastOp>(
+        loc, LLVM::LLVMPointerType::get(elemTy, resTy.getMemorySpaceAsInt()),
+        ptr);
+    ptr = rewriter.create<Pointer2MemrefOp>(loc, resTy, ptr);
+
+    rewriter.replaceOp(op, {ptr});
+    return success();
```

Comment:
```suggestion
    const auto newTy = getTypeConverter()->convertType(op.getType());                                                                                                                                                                                                                                                        
    rewriter.replaceOpWithNewOp<LLVM::AddrSpaceCastOp>(op, newTy, 
      transformed.getSource());
   return success();
```
`transformed` already contains the transformed input, no need to `Memref2PointerOp`. Also, no need to `Pointer2MemrefOp` because an `LLVMPointer` of `newTy` is expected.